<!doctype html>
<html><head>
<meta charset="utf-8">
<title>WAT knowledge portal</title>
<link rel="icon" type="image/png" href="images/favicon.png">
<link href="style.css" rel="stylesheet" type="text/css">
<style>
  @import url('https://fonts.googleapis.com/css2?family=Red+Hat+Display:wght@300&display=swap');
</style>
<!-- basic -->
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1">
<!-- mobile metas -->
<meta name="viewport" content="width=device-width, initial-scale=1">
<meta name="viewport" content="initial-scale=1, maximum-scale=1">
<!-- site metas -->
<title>WATkp</title>
<meta name="keywords" content="">
<meta name="description" content="">
<meta name="author" content="">	
</head>

<body class="container">
<header>
    <nav class="navbar_in">
	    <a href="index.html"><img src="images/logo_navbar.png" alt="adipose tissue knowledge portal logo" class="logo"></a> 
		
<div class="navbar">
			 
  <a href="index.html">Home</a>
  <div class="dropdown">
    <button class="dropbtn">Modules
      <i class="fa fa-caret-down"></i>
    </button>
    <div class="dropdown-content">
	  <a href="gene_sum.html">Summary</a>
      <a href="clinical.html">Clinical relationships</a>
      <a href="depots.html">Depot comparisons</a>
      <a href="celltype.html">Cell type</a>
	  <a href="sc.html">Single-cell/single-nucleus RNA seq.</a>
      <a href="spatial.html">Spatial Transcriptomics</a>
      <a href="pert.html">Perturbation</a>
	  <a href="gene_query.html">Gene finder</a>
    </div>
  </div>
  
   <div class="dropdown">
    <button class="dropbtn">About
      <i class="fa fa-caret-down"></i>
    </button>
    <div class="dropdown-content">
      <a href="vignettes.html">How to</a>
	  <a href="news.html">News</a>
	  <a href="contact.html">Contact</a>
      <a href="version_updates.html">Version Updates</a>
    </div>
  </div>
</div> 
	</nav>
  </header>
	
  <main>
   
    <div id="main"> 
  	  <div id="sidebar">
		<div class="sidebar_st">
		  <div class="text_sidebar">
			  <div><p>Introduction</p></div>
			     <div><p>How to</p></div>
			     <div><p>Clinical cohorts</p></div>
                 <div><p>Cell type datasets</p></div>
          </div>
        </div>
    </div>
	
	  <div id="content" class="vign_articles">
		  
		  	<div class="contact_text" id="intro">  <h2>Introduction</h2> </div>
		  
		    <div> <hr class="separation_vign"> </div>
		  
		    <div class="contact_text" id="intro_text">
				
		    <p>Text</p>
		  
		    </div>
	  
		  
	  
	 <div class="contact_text">
	    <p>
        </p>
     </div>
   
<div class="contact_text"><h2>How to</h2>  </div>
	  
	  <div> <hr class="separation_vign" > </div>
	  
	 <div class="contact_text">
		 <p> Text
        </p>
     </div>
		  
<div class="contact_text"><h2>Clinical cohorts</h2>  </div>
	  
	 <div> <hr class="separation_vign" > </div>
	  
	 <div class="contact_text">
		 
	    <p><b>Krieg, L. et al. (2021)</b></br>
		For this cohort, the authors collected subcutaneous, omental, mesenteric, and epiploic adipose tissue from a total of 70 insulin resistant and insulin sensitive subjects with overweight and obesity undergoing bariatric surgery. Patients were recruited at the University Hospital Leipzig, Germany. Transcriptome data was generated by microarray (HT12 v4) for 24 subjects.</br>
		Publication: <i>“Multiomics reveal unique signatures of human epiploic adipose tissue related to systemic insulin resistance”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/34598978/" target="_blank"> (PMID: 34598978)</a></p>
	  
	    <p><b>Keller, M. et al. (2017)</b></br>
		This cohort focused on differences in DNA methylation between subcutaneous and omental adipose tissue, which was analyzed along mRNA profiles of paired samples from 105 subjects undergoing abdominal surgery. Patients were recruited at the University Hospital Leipzig, Germany. DNA promoter methylation (GeneChip Human Promoter 1.0R Arrays) is available for 77 subjects, while for 63 subjects transcriptome data (HT-12) is available.</br>
		Publication: <i>“Genome-wide DNA promoter methylation and transcriptome analysis in human adipose tissue unravels novel candidate genes for obesity”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/28123940/" target="_blank"> (PMID: 28123940)</a></p>
	    <p><b>Kerr, A. et al. (2020) </b></br>
		PENDING</br>
		Publication: <i>“Long-term changes in adipose tissue gene expression following bariatric surgery”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/32406570/" target="_blank"> (PMID: 32406570)</a></p>

	    <p><b>Imbert, A. et al. (2022)</b></br>
		Imbert et al. sequenced subcutaneous white adipose tissue samples of 556 individuals that were enrolled in an 8-week low-caloric weight loss diet study (DiOGenes) using QuantSeq sequencing. The DiOGenes study is a pan-European, multicenter study that was originally described by Larsen et al. (N Engl J Med., 2010;363(22): 2102-2113.). Samples were collected from the abdomen by needle aspiration from about 10 cm from the umbilicus.</br>
		Publication: <i>“Network Analyses Reveal Negative Link Between Changes in Adipose Tissue GDF15 and BMI During Dietary-induced Weight Loss”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/34415992/" target="_blank"> (PMID: 34415992)</a></br>
        Clinical trial registration: 
        <a href="https://clinicaltrials.gov/study/NCT00390637?term=NCT00390637&rank=1" target="_blank"> NCT00390637</a></p>

	    <p><b>Armenise, C. et al. (2017)</b></br>
		This publication is part of the DiOGenes cohort, a pan-European, multicenter, randomized controlled dietary intervention program (Larsen et al., N Engl J Med., 2010;363(22): 2102-2113.). After quality-filtering, the authors analyzed transcriptome data of 191 subjects with obesity but without diabetes.</br>
		Publication: <i>“Genome-wide DNA promoter methylation and transcriptome analysis in human adipose tissue unravels novel candidate genes for obesity”</i>
        <a href="https://pubmed.ncbi.nlm.nih.gov/28793995/" target="_blank"> (PMID: 28793995)</a></br>
        Clinical trial registration: 
        <a href="https://clinicaltrials.gov/study/NCT00390637?term=NCT00390637&rank=1" target="_blank"> NCT00390637</a></p>

	    <p><b>Arner, P. et al. (2016)</b></br>
		This publication is part of the EMIF-Metabolic study, an IMI-funded project aiming to identify novel biomarkers of obesity complications. Here, transcriptomic and epigenetic data are available from subcutaneous abdominal and visceral (omental) white adipose tissue of 80 women living with obesity. Participants were subdivided into insulin resistant and insulin sensitive clinical phenotypes (measured by HOMA-IR), and were recruited from a larger cohort of 220 women who participated in a clinical trial on the effects of bariatric surgery .</br>
		Publication: <i>“The epigenetic signature of systemic insulin resistance in obese women”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/27535281/" target="_blank"> (PMID: 27535281)</a></br>
        Clinical trial registration: 
        <a href="https://clinicaltrials.gov/study/NCT01785134?term=NCT01785134&rank=1" target="_blank"> NCT01785134</a></p>

	    <p><b>Petrus, P. et al. (2018)</b></br>
		This publication includes 16 women living with obesity investigated before and two years after bariatric surgery. All women attained a post-operative body mass index lower than 30 kg/m2 and for each post-obese subject, a woman who had never been obese was recruited and matched for age and current BMI. From all participants, subcutaneous abdominal white adipose tissue biopsies were obtained and investigated by gene microarray.</br>
		Publication: <i>“Transforming Growth Factor-β3 Regulates Adipocyte Number in Subcutaneous White Adipose Tissue”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/30332637/" target="_blank"> (PMID: 30332637)</a></p>

	    <p><b>Arner, E. et al. (2012)</b></br>
		In this publication, 56 women were recruited and subdivided in 26 non-obese and 30 individuals living with obesity. Subcutaneous abdominal white adipose tissue biopsies were obtained and analyzed by gene microarray.</br>
		Publication: <i>“Adipose tissue microRNAs as regulators of CCL2 production in human obesity”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/22688341/" target="_blank"> (PMID: 22688341)</a></p>

	    <p><b>Civelek, M. et al. (2017)</b></br>
		The study by Civelek et al. is part of the population-based METSIM study (Stancáková et al., Diabetes, 58 (2009), pp. 1212-1221). The authors performed expression quantitative trait locus analyses in abdominal subcutaneous adipose tissue samples of 770 male subjects. Subjects were recruited and underwent phenotyping at the Clinical Research Unit at the University of Kuopio, Finland. Transcriptome data was obtained using the Affymetrix U219 microarray.</br>
		Publication: <i>“Genetic Regulation of Adipose Gene Expression and Cardio-Metabolic Traits”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/28257690/" target="_blank"> (PMID: 28257690)</a></p>

	    <p><b>Raulerson, C. et al. (2019)</b></br>
		Raulerson et al. analyzed RNA sequencing data of subcutaneous adipose tissue from 434 male subjects. All individuals were recruited as part of the METSIM study (Stancáková et al., Diabetes, 58 (2009), pp. 1212-1221). Subjects were recruited and underwent phenotyping at the Clinical Research Unit at the University of Kuopio, Finland. RNA sequencing was performed using the Illumina TruSeq RNA kit v2 on an Illumina HiSeq 2000 platform.</br>
		Publication: <i>“Adipose Tissue Gene Expression Associations Reveal Hundreds of Candidate Genes for Cardiometabolic Traits”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/31564431/" target="_blank"> (PMID: 31564431)</a></p>

	    <p><b>Keller, M. et al. (2017)</b></br>
		This cohort focused on differences in DNA methylation between subcutaneous and omental adipose tissue, which was analyzed along mRNA profiles of paired samples from 105 subjects undergoing abdominal surgery. Patients were recruited at the University Hospital Leipzig, Germany. DNA promoter methylation (GeneChip Human Promoter 1.0R Arrays) is available for 77 subjects, while for 63 subjects transcriptome data (HT-12) is available.</br>
		Publication: <i>“Genome-wide DNA promoter methylation and transcriptome analysis in human adipose tissue unravels novel candidate genes for obesity”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/28123940/" target="_blank"> (PMID: 28123940)</a></p>

	    <p><b>Stančáková, A. (2012)</b></br>
		This cohort is part of the large METSIM study (Stancáková et al., Diabetes, 58 (2009), pp. 1212-1221), which is a population-based Finish cohort. Subjects were recruited and underwent phenotyping at the Clinical Research Unit at the University of Kuopio, Finland. The authors generated an mRNA expression data set from 200 male subjects using the Illumina Human HT-12 v3 Expression BeadChip.</br>
		Publication: <i>“Hyperglycemia and a Common Variant of GCKR Are Associated With the Levels of Eight Amino Acids in 9,369 Finnish Men”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/22553379/" target="_blank"> (PMID: 22553379)</a></p>

	    <p><b>Winnier, DA. (2015)</b></br>
		Winnier et al. analyzed the transcription profile of subcutaneous adipose tissue in 308 Mexican Americans who were enrolled in the Veterans Administration Genetic Epidemiology Study (VAGES) and compared it to 75 unrelated control individuals. Gene expression was measured using the Affymetrix HG-U133A microarray.</br>
		Publication: <i>“Transcriptomic identification of ADH1B as a novel candidate gene for obesity and insulin resistance in human adipose tissue in Mexican Americans from the Veterans Administration Genetic Epidemiology Study (VAGES)”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/25830378/" target="_blank"> (PMID: 25830378)</a></p>

	    <p><b>Nookaew, I. (2013)</b></br>
		In this study, which is part of the SOS Sib Pair study, the author compared adipose tissue transcriptome profiles between 113 men and 262 women. Microarray profiles from perithyroid and paired subcutaneous adipose tissues were generated using the Human Genome U133 plus 2.0 or U133A DNA microarrays.</br>
		Publication: <i>“Adipose tissue resting energy expenditure and expression of genes involved in mitochondrial function are higher in women than in men”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/23264395/" target="_blank"> (PMID: 23264395)</a></p>

	    <p><b>Keller, P. (2011)</b></br>
		Keller et al. studied the impact of microRNAs on adipogenesis and obesity. For this effort they collected, among other things, 20 samples of subcutaneous adipose tissue from subjects with a BMI range of 21-48 kg/m2. The expression was measured using U133+2 Affymetrix arrays.</br>
		Publication: <i>“Gene-chip studies of adipogenesis-regulated microRNAs in mouse primary adipocytes and human obesity”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/21426570/" target="_blank"> (PMID: 21426570)</a></p>

	    <p><b>Das, SK. (2015)</b></br>
		In this comparative study on Americans with either African or Caucasian descent, Das et al. collected subcutaneous from 136 individual. Subjects had no diabetes but a wide range of insulin sensitivity and BMI. Expression profile was assessed using the HumanHT-12 v4 Expression BeadChip.</br>
		Publication: <i>“Integrative network analysis reveals different pathophysiological mechanisms of insulin resistance among Caucasians and African Americans”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/25868721/" target="_blank"> (PMID: 25868721)</a></p>

	    <p><b>Barberio, MD. (2019)</b></br>
		Barberio et al. analyzed epigenetic changes in visceral adipose tissue and therefore collected specimens from 30 adolescent women, of which 15 were lean and 15 obese. The global expression of visceral omental adipose tissue was assessed using Affymetrix Hu133 Plus 2.0 microarrays.</br>
		Publication: <i>“Comparison of visceral adipose tissue DNA methylation and gene expression profiles in female adolescents with obesity”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/31798691/" target="_blank"> (PMID: 31798691)</a></p>

	    <p><b>Vink, RG. (2017)</b></br>
		For this study on moderate weight loss using a low-caloric (1250 kcal per day for 12 weeks) or very low-caloric diet (500 kcal per day for five weeks), the authors collected abdominal subcutaneous adipose tissue samples from 25 male and 28 female subjects. Transcriptome data was obtained using human whole-genome Affymetrix Gene 1.1 ST arrays.</br>
		Publication: <i>“Adipose tissue gene expression is differentially regulated with different rates of weight loss in overweight and obese humans”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/27840413/" target="_blank"> (PMID: 27840413)</a></p>

	    <p><b>Sharma, NK. (2016)</b></br>
		In this cohort of African Americans, the authors obtained transcriptome data of 260 individuals using the HumanHT-12 v4 Expression BeadChip. All studies were performed at the Wake Forest School of Medicine (WFSM) Clinical Research Unit. Participants were apparently healthy, between 18 and 60 years of age and had a BMI between 18 and 42 kg/m<sup>2</sup>.</br>
		Publication: <i>“Genome-wide DNA promoter methylation and transcriptome analysis in human adipose tissue unravels novel candidate genes for obesity”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/26789776/" target="_blank"> (PMID: 26789776)</a></p>

	    <p><b>GTEx microarray Lonsdale, J. (2013)????</b></br>
		PENDING.</br>
		Publication: <i>“The Genotype-Tissue Expression (GTEx) project”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/23715323/" target="_blank"> (PMID: 23715323)</a></p>

	    <p><b>Aguet, F. (2017)</b></br>
		This cohort is part of the Genotype-Tissue Expression project (GTEx) v6. For this large project, the authors collected postmortem samples of 449 human donors across 44 tissues. Here, we included XXX samples with available data for subcutaneous and omental adipose tissue. RNA sequencing data was prepared according to the Illumina TruSeq protocol.</br>
		Publication: <i>“Genetic effects on gene expression across human tissues”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/29022597/" target="_blank"> (PMID: 29022597)</a></p>

	    <p><b>GTEx_v7</b></br>
		This cohort focused on .......</br>
		Publication: <i>“.....”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/28123940/" target="_blank"> (PMID: .....)</a></p>

	    <p><b>Aguet, F. (2020)</b></br>
		This adipose tissue transcriptome data was taken from the Genotype-Tissue Expression project (GTEx) v8. The authors analyzed 15,201 samples comprising 49 tissues of 838 postmortem donors. Here, we included XXX samples with available data for subcutaneous and omental adipose tissue.</br>
		Publication: <i>“The GTEx Consortium atlas of genetic regulatory effects across human tissues”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/32913098/" target="_blank"> (PMID: 32913098)</a></p>

	    <p><b>Drong, A. (2013)</b></br>
		For this study, abdominal subcutaneous fat samples have been collected from the Oxford Biobank (Tan, G. D. et al., Diabetologia 49: 158–168). 19 samples with metabolic syndrome and 19 controls samples from 18 male and 20 female subjects have been included in the study and mRNA expression was measured using the Affymetrix human GeneChip HGU133 Plus 2.0 array.</br>
		Publication: <i>“The presence of methylation quantitative trait loci indicates a direct genetic influence on the level of DNA methylation in adipose tissue”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/23431366/" target="_blank"> (PMID: 23431366)</a></p>

	    <p><b>Naukkarinen, J. (2013)</b></br>
		Naukkarinen et al. studied subcutaneous adipose tissue expression of 16 obesity-discordant twin pairs using Affymetrix U133 Plus 2.0 chips. Subjects have been recruited from the FinnTwin16 and FinnTwin12 studies and descriptive details have been published prior by Graner et al. (Am J Cardiol 109:1295–1302).</br>
		Publication: <i>“Characterising metabolically healthy obesity in weight-discordant monozygotic twins”</i>
		<a href="https://pubmed.ncbi.nlm.nih.gov/24100782/" target="_blank"> (PMID: 24100782)</a></p>

	    
     </div>

<div class="contact_text">  <h2>Cell type datasets</h2>  </div>
	  
	  <div> <hr class="separation_vign" > </div>

        <div class="contact_text">
		 
	    <p><b>Adipogenesis - adipose tissue derived stem cells</b></br>
		This data set allows a detailed look at effects of adipogenesis on transcript levels in adipose tissue derived stem cells. Details on the differentiation are reported under the “Details” section. Data is available for 16 timepoints between induction of adipogenesis and the differentiated state (day 12). Data on the x axis is presented as natural logarithm of the minutes after induction of adipogenesis. Data was generated at Ryden & Mejhert lab.</br>
	  
	    <p><b>Adipogenesis - SVF and isolated adipocytes</b></br>
		For this data set on ex vivo differentiation of isolated fibroblasts the stroma vascular fraction of abdominal subcutaneous adipose tissue from four individuals was obtained by collagenase digestion. Cells were differentiated as detailed in the “Details” section, and RNA was collected at 4, 8 and 12 days after differentiation. Transcript information was obtained by CAGE sequencing. Isolated adipocytes from the same subjects serve as control samples. </br>

	    <p><b>FACS Fractionation</b></br>
		In this data set Acosta et al. isolated subcutaneous adipose tissue from 6 women (see Details) and performed collagenase digest to separate adipocytes and stroma vascular fraction (SVF). Subsequent FACS sorting allowed the generation of adipose tissue and cell type specific gene expression data for major cell classes including T cells, M1 and M2 macrophages, endothelial cells using Clariom D microarrays.</br>

	  </div>
	  
	 <div class="contact_text">
		 <p>
        </p>
     </div>
		  
	  </div>
    </div>
  </main>

<footer id="footer">
	 <div class="footer_col">
		 <div class="grid-container_bot">
		 <div class="grid-child">
        <h4>Publication</h4>
                  <p> The WATportal manuscript is available via bioRxiv [link to bioRxiv] </p>
    </div>
			 <div class="grid-child">
        <h4>Data and code:</h4>
		          	<p>The code for this portal is available through <a href="https://github.com/jiawei-zhong/WAT_portal"> github</a>.</p>
    </div>
			 <div class="grid-child">
        <h4>Contribute</h4>
		          	<p>Check the <a href="contact.html">About</a> page to learn how to contribute to the WATportal.</p>
    </div>
			 <div class="grid-child">
        <h4>Contact</h4>
		          	<p>Contact details are provided on the <a href="contact.html">About</a> page.</p>
    </div>
			 <div class="grid-child">
        <h4>Version update</h4>
		          	<p>v1.0.0. (updated: 01 Nov. 2023)</p>
    </div>
		 </div>
		 
     </div> 
</footer>

</body>
</html>
